docs/citycloud-snippets.md

```diff
@@ -0,0 +1,130 @@
+      -e "s@^image = .*@image = \"${OS_IMAGE_NAME}\"@" \
+      -e "s@^ssh_user = .*@ssh_user = \"ubuntu\"@" \
+      -e "s@^number_of_k8s_masters = .*@number_of_k8s_masters = ${OS_NUM_MASTERS}@" \
+      -e "s@^number_of_k8s_nodes = .*@number_of_k8s_nodes = ${OS_NUM_WORKERS}@" \
+      -e "s@^number_of_k8s_nodes_no_floating_ip = .*@number_of_k8s_nodes_no_floating_ip = 0@" \
+      -e "s@^flavor_k8s_master = .*@flavor_k8s_master = \"${OS_MASTERS_FLAVOR}\"@" \
+      -e "s@^#\?flavor_k8s_node = .*@flavor_k8s_node = \"${OS_WORKER_FLAVOR}\"@" \
+      -e "s@^network_name = .*@network_name = \"${CLUSTER}-network\"@" \
+      -e "s@^external_net = .*@external_net = \"${OS_EXT_NET_ID}\"@" \
+      -e "s@^subnet_cidr = .*@subnet_cidr = \"${OS_SUBNET_CIDR}\"@" \
+      -e "s@^floatingip_pool = .*@floatingip_pool = \"${OS_FLOATINGIP_POOL}\"@" \
+  < "contrib/terraform/openstack/sample-inventory/cluster.tfvars" > "inventory/$CLUSTER/cluster.tfvars"
+done
+popd
+```
+
+Check the `cluster.tfvars` file and make sure your settings are what you want.
+If you want to be able to ssh into the machines you need to set the `k8s_allowed_remote_ips` variable to something like:
+
+```tfvars
+k8s_allowed_remote_ips = ["1.2.3.4/32"]
+```
+
+Now you're ready to deploy the infrastructure
+
+```bash
+for CLUSTER in "${SERVICE_CLUSTER}" "${WORKLOAD_CLUSTERS[@]}"; do
+  pushd "kubespray/inventory/${CLUSTER}"
+  terraform init ../../contrib/terraform/openstack
+  terraform apply \
+    -var-file cluster.tfvars \
+    -auto-approve \
+    -state="tfstate-${CLUSTER}.tfstate" \
+    ../../contrib/terraform/openstack
+  popd
+done
+```
+
+If all went well the infrastructure should now be up and running
+
+## Compliantkubernetes kubespray
+
+To set up kubernetes with compliantkubernetes-kubespray you can follow these steps.
+
+Export your cluster path and PGP fingerprint
+
+```bash
+export CK8S_CONFIG_PATH=~/.ck8s/my-new-cluster
+export CK8S_PGP_FP=<Your PGP fingerprint>
+```
+
+Initialize the configuration with.
+
+```bash
+for CLUSTER in "${SERVICE_CLUSTER}" "${WORKLOAD_CLUSTERS[@]}"; do
+  ./bin/ck8s-kubespray init "${CLUSTER}" citycloud ~/.ssh/id_rsa
+  ln -s "$(pwd)/kubespray/inventory/${CLUSTER}/tfstate-${CLUSTER}.tfstate" "${CK8S_CONFIG_PATH}/${CLUSTER}-config/" || true
+  cp "kubespray/contrib/terraform/openstack/hosts" "${CK8S_CONFIG_PATH}/${CLUSTER}-config/inventory.ini"
+  chmod +x "${CK8S_CONFIG_PATH}/${CLUSTER}-config/inventory.ini"
+done
+```
+
+Check the variables in the `group_vars` folder for each cluster and make sure that it matches what you want.
+
+Apply the configuration and set up kubernetes.
+
+```bash
+for CLUSTER in "${SERVICE_CLUSTER}" "${WORKLOAD_CLUSTERS[@]}"; do
+  TERRAFORM_STATE_ROOT=${CK8S_CONFIG_PATH}/${CLUSTER}-config/ bin/ck8s-kubespray apply "${CLUSTER}"
+done
+```
+
+Done!
+
+## Teardown
+
+Later when you want to destroy the infrastructure.
+
+Make sure all cloud resources are destroyed (persistent volumes, load balancers).
```

Comment:
Is that really that dramatic? At least in AWS, the resources created by Kubernetes are clearly tagged and can easily be removed manually. Maybe we should find the snipped that does so?

Reply:
I don't think it is that easy to find them in other clouds but I could be wrong. But more importantly, this could cause a surprise bill for people who aren't aware of it. If you know terraform but not kubernetes, it wouldn't be obvious at all that there can be resources left in the cloud after you do `terraform destroy`.

Relevant issue: https://github.com/elastisys/compliantkubernetes-apps/issues/213

Reply:
I see two aspects you bring here:

1. Regarding the Operator Manual, it must start from the assumption that people have CKA or equivalent under the bell. That curriculum does a way better job at explaining the ins-and-outs of Kubernetes than we will ever do. We should focus our documentation on what Compliant Kubernetes *adds* on top of a vanilla Kubernetes cluster, i.e., on what differentiates us, and not redo a half-finished, incomplete, out-of-date version of CKA. :smile: 
2. If the interaction between Kubernetes and underlying cloud resources is not well known or well understood in the community, this would be an excellent blog post.